packages/cli/src/fulltext/querySymbols.ts

```diff
@@ -0,0 +1,78 @@
+import queryKeywords from './queryKeywords';
+import { readFileSync } from 'fs';
+
+export const SymbolRegexes: Record<string, RegExp[]> = {
+  cs: [
+    /(interface|class|enum|struct)\s+(?<symbol>\w+)/g,
+    /(\s|^)(?!using|try|catch|if|while|do|for|switch)(?<!#define\s+?)(?<symbol>[\w~$]+)\s*?\([^;)]*?\)[\w\s\d<>[\].:\n]*?{/g,
+  ],
+  cpp: [
+    /(struct|enum|union|class)\s+(?<symbol>\w+)\s*?\{/g,
+    /}\s*?(?<symbol>\w+)\s*?;/g,
+    /(\s|^)(?!try|catch|if|while|do|for|switch)(?<!#define\s+?)(?<symbol>[\w~$]+)\s*?\([^;)]*?\)[\w\s\d<>[\].:\n]*?{/g,
+  ],
+  rs: [/(struct|enum|union|trait|type|fn)\s+(?<symbol>[\w\p{L}]+)/gu],
+  go: [/type\s+(?<symbol>[\w\p{L}]+)/gu, /func\s+?(\(.*?\)\s*?)?(?<symbol>[\w\p{L}]+)\s*?\(/gu],
+  rb: [/(class|module)\s+(?<symbol>\w+)/g, /def\s+?(?<symbol>\w+)/g],
+  py: [/(class|def)\s+(?<symbol>\w+)/g],
+  java: [
+    /(class|@?interface|enum)\s+(?<symbol>[\w$]+)/g,
+    /(\s|^)(?!try|catch|if|while|do|for|switch)(?<symbol>[\w$]+)\s*?\([^;)]*?\)[\w\s\d<>[\].:\n]*?{/g,
+  ],
+  ts: [
+    /(class|interface|enum|type|function)\s+(?<symbol>[#$\w\p{L}]+)/gu,
+    /(\s|^)(?!using|try|catch|if|while|do|for|switch)(?<symbol>[#$\w\p{L}]+)\s*?\([^;)]*?\)[\w\s<>[\].:\n]*?\{/gu,
+    /(?<symbol>[#$\w\p{L}]+)\s*?(=|:)\s*?\(.*?\)\s*?=>/gu,
+  ],
+  kt: [/(class|typealias)\s+(?<symbol>[\w_]+)/g, /fun\s+?(<.+?>\s+)?(.*?\.)?(?<symbol>\w+)/g],
+  php: [/(class|trait|function)\s+(?<symbol>[\w\d_$]+)/g],
+};
+
+const genericSymbolRegexes: RegExp[] = [
+  /(^|\s)(?!using|try|catch|if|while|do|for|switch)(?<symbol>[#$\w\p{L}~]+)\s*?\(([^;)])*?\)[\w\s<>[\].:\n]*?\{/gmu,
+  /^(?!.*?(?:#|\/\/|"|')).*?(interface|class|enum|struct|union|trait|type(alias|def)?|fu?nc?(tion)?|module|def)\s+?(?<symbol>[#$\w\p{L}]+)/gmu,
+  /(?<symbol>[#$\w\p{L}~]+)\s*?=\s*?[\w\s<>[\].:\n]*?\{/gu,
+];
+
+// Define aliases for common file extensions
+['js', 'jsx', 'ts', 'tsx', 'vue', 'svelte'].forEach((ext) => {
+  SymbolRegexes[ext] = SymbolRegexes.ts;
+});
+
+['c', 'cc', 'cxx', 'h', 'hpp', 'cpp', 'hxx'].forEach((ext) => {
+  SymbolRegexes[ext] = SymbolRegexes.cpp;
+});
+
+function getUniqueMatches(source: string, regex: RegExp[]): string[] {
+  const matches = new Set(
+    regex
+      .flatMap((re) => Array.from(source.matchAll(re)))
+      .filter(Boolean)
+      .map(({ groups }) => groups?.symbol ?? '')
+  );
+  return Array.from(matches).sort();
+}
+
+export default function querySymbols(path: string, applyKeywordTransform = true): string[] {
+  const fileExtension = path.split('.').pop()?.toLowerCase();
+  let regexes = genericSymbolRegexes;
+  if (fileExtension && fileExtension in SymbolRegexes) {
+    regexes = SymbolRegexes[fileExtension];
+  }
+
+  let result: string[] = [];
+  try {
+    const content = readFileSync(path, 'utf-8');
```

Comment:
Why `readFileSync`?

Reply:
The async method caused indexing to run 2x slower overall. The parallel I/O seems to introduce more overhead than it's worth in this case, which isn't too surprising given we're reading files one after another anyway.

Reply:
Ok I think I see what you mean. I thought you changed the loop structure from a Promise.all to a for loop. Because with a for loop the access to the different files will not be interwoven, but it can also be non blocking,

For example in the RPC server there might be other calls running concurrently (like a context lookup or loading AppMap data) and with the sync read those operations would be blocked.